File: src/jvmMain/kotlin/Inheritance.kt
fun main() {
    println("Kotlin console below\n")
    Derived("Mars").run {
        questionName()
        sayHello()
        sayBye()
    }
}

open class Base(val name: String) {
    open fun sayHello() {
        println("Hello $name!")
    }
}

fun Base.questionName() {
    println("You're $name, right?")
}

class Derived(private val derivedName: String) : Base("definitely not $derivedName") {
    fun sayBye() {
        println("Bye-bye $name!")
    }

    override fun sayHello() {
        println("What's up $derivedName!")
    }
}

File: src/Inheritance.js
class Base {
    constructor(name) {
        this.name = name
    }

    sayHello() {
        console.log(`Hello ${this.name}!`)
        return this
    }
}

Base.prototype.questionName = function () {
    console.log(`You're ${this.name}, right?`)
    return this
}

class Derived extends Base {
    constructor(derivedName) {
        super(`definitely not ${derivedName}`)
        this.derivedName = derivedName
    }

    sayBye() {
        console.log(`Bye-bye ${this.name}!`)
        return this
    }

    sayHello() {
        console.log(`What's up ${this.derivedName}!`)
        return this
    }
}

console.debug("JS console below\n")
new Derived("Mars")
    .questionName()
    .sayHello()
    .sayBye()